bentoml/yatai/grpc_stream_utils.py

```diff
@@ -0,0 +1,138 @@
+import os
+import io
+import tarfile
+from bentoml.yatai.proto.repository_pb2 import UploadBentoRequest, DownloadBentoResponse
+
+
+DEFAULT_FILE_CHUNK_SIZE = 1024 * 8
+
+
+class FileStream:
+    def __init__(self):
+        self.buffer = io.BytesIO()
+        self.offset = 0
+
+    def write(self, s):
+        self.buffer.write(s)
+        self.offset += len(s)
+
+    def tell(self):
+        return self.offset
+
+    def close(self):
+        self.buffer.close()
+
+    def read_value(self):
+        try:
+            return self.buffer.getvalue()
+        finally:
+            self.buffer.close()
+            self.buffer = io.BytesIO()
+
+
+class BentoBundleStreamRequestsOrResponses:
+    def __init__(
+        self,
+        bento_name,
+        bento_version,
+        directory_path,
+        file_chunk_size=DEFAULT_FILE_CHUNK_SIZE,
+        is_request=True,
+    ):
+        """
+        Build a tar file chunk by chunk on the fly from a directory and then stream out
+        either request or response for upload/download.
+
+        Args:
+            bento_name:
+            bento_version:
+            directory_path:
+            file_chunk_size:
+            is_request:
+        """
+        self.bento_name = bento_name
+        self.bento_version = bento_version
+        self.directory_path = directory_path
+        self.file_chunk_size = file_chunk_size
+        self.is_request = is_request
+        self.out_stream = FileStream()
+        self.tar = tarfile.TarFile(fileobj=self.out_stream, mode='w')
+
+    @staticmethod
+    def _stream_file_into_tar(tarinfo, tar, file_handler, buf_size):
+        out = tar.fileobj
+
+        # For iter(), the second argument, sentinel, is given, then object must
+        # be a callable object. The iterator created in this case will
+        # call object with no arguments for each call to its __next__()
+        # method; if the value returned is equal to sentinel, StopIteration
+        # will be raised, otherwise the value will be returned.
+        for b in iter(lambda: file_handler.read(buf_size), b''):
+            out.write(b)
+            yield
+        blocks, remainder = divmod(tarinfo.size, tarfile.BLOCKSIZE)
+        if remainder > 0:
+            out.write(tarfile.NUL * (tarfile.BLOCKSIZE - remainder))
+            blocks += 1
+        tar.offset += blocks * tarfile.BLOCKSIZE
+        yield
+
+    def create_request_or_response(self):
+        if self.is_request:
+            return UploadBentoRequest(
+                bento_name=self.bento_name,
+                bento_version=self.bento_version,
+                bento_bundle=self.out_stream.read_value(),
+            )
+        else:
+            return DownloadBentoResponse(bento_bundle=self.out_stream.read_value())
+
+    def close(self):
+        self.tar.close()
+        self.out_stream.close()
+
+    def __iter__(self):
```

Comment:
could we add some standalone tests to verify this behavior?  e.g. have a test case that generates random files under nested directories and verifies the received file has identical structure and content?